client/services/code_evaluators/PythonPrerequisiteCheckService.js

```diff
@@ -0,0 +1,126 @@
+ */
+
+tie.factory('PythonPrerequisiteCheckService', [
+  'PYTHON_STANDARD_LIBRARIES',
+  'PrerequisiteEvalResultObjectFactory',
+  function(
+      PYTHON_STANDARD_LIBRARIES,
+      PrerequisiteEvalResultObjectFactory) {
+
+    const LEFT_TRIM_PATTERN = /^\s+/;
+
+    var checkStarterCodePresent = function(starterCode, code) {
+      var starterCodeLines = starterCode.split('\n');
+      var codeLines = code.split('\n');
+      for (var i = 0; i < codeLines.length; i++) {
+        codeLines[i] = codeLines[i].trim();
+      }
+      var codeLineIndex = 0;
+      for (var i = 0; i < starterCodeLines.length; i++) {
+        var starterCodeLine = starterCodeLines[i].trim();
+        // ignore blank lines in starter code
+        if (starterCodeLine.length === 0) {
+          continue;
+        }
+        //handle "return" statements in starter code
+        if (starterCodeLine.replace(LEFT_TRIM_PATTERN,"").startsWith("return ")) {
+          while (codeLineIndex < codeLines.length) {
+            if (codeLines[codeLineIndex].replace(
+              /^\s+/,"").startsWith("return ")) {
+                break;
+            }
+            codeLineIndex++;
+          }
+        }
+        else {
+          while (codeLineIndex < codeLines.length) {
+            if (codeLines[codeLineIndex] === starterCodeLine) {
+              break;
+            }
+            codeLineIndex++;
+          }
+        }
+        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    var getCodeLibs = function(code) {
+      var codeLines = code.split('\n');
+      var coddLibs = [];
+      var pattern = new RegExp('^import\\ \\w+$');
+      for(var i=0; i<codeLines.length; i++) {
+        if(pattern.test(codeLines[i])) {
+          var words = codeLines[i].split(' ');
+          coddLibs.push(words[1]);
+        }
+      }
+      return coddLibs;
+    };
+
+    return {
+      // Returns a promise.
+      checkCode: function(starterCode, code) {
+        if (!(checkStarterCodePresent(starterCode, code))) {
+          var errorMessage = ['It looks like you deleted or modified ',
+            'the starter code!  Our evaluation program requires the ',
+            'function names given in the starter code.  You can press the ',
+            '\'Reset Code\' button below to start over.  Or, you can copy ',
+            'the starter code below:'].join('');
+          return Promise.resolve(
+            PrerequisiteEvalResultObjectFactory.create(
+              code, errorMessage, starterCode));
+        }
+
+        var libErrorMessage = ['It looks like you are importing an external ',
+                        'library.  Only standard libraries are supported.  ',
+                        'The following libraries are not ',
+                        'supported:\n'].join('');
+        codeLibs = getCodeLibs(code);
+        var extLibs = [];
```

Comment:
Again, all this stuff is way too complicated. You could instead do something like this:

    var unsupportedImports = importedLibraries.filter(function(importedLibrary) {
      return PYTHON_STANDARD_LIBRARIES.indexOf(importedLibrary) === -1;
    });

Make sure that the importedLibraries list is already trimmed, etc. Then pass the unsupportedImports as a customization to the prereq check result factory.